lib/i18n/backend/lazy_load.rb

```diff
@@ -0,0 +1,92 @@
+    # many different locales, but only refer to a single locales' translations
+    # at a time (ex. A Rails workload).  The implementation identifies which
+    # translation files from the load path belong to the current locale by
+    # pattern matching against their path name.
+    #
+    # Specifically, a translation file is considered to belong to a locale if:
+    # a) the filename is in the I18n load path
+    # b) the filename ends in a supported extension (ie. .yml, .json, .po, .rb)
+    # c1) the filename starts with the locale identifier
+    #    OR
+    # c2) the path contains a component equal to the locale identifier
+    #
+    # Examples:
+    # Valid files that will be selected by this backend:
+    #
+    # "files/locales/en-translation.yml" (Selected for locale "en")
+    # "files/locales/fr/translation.po"  (Selected for locale "fr")
+    #
+    # Invalid files that won't be selected by this backend:
+    #
+    # "files/locales/translation-file"
+    # "files/locales/en-translation.unsupported"
+    # "files/locales/french/translation.yml"
+    #
+    # The implementation uses this assumption to defer the loading of
+    # translation files until the current locale actually requires them.
+    #
+    # To use the LazyLoad backend instantiate it and set it to the I18n module.
+    # You can configure lazy loaded backends through the initializer or backends
+    # accessor:
+    #
+    #   # configure I18n backend to use LazyLoad Backend
+    #   I18n.backend = I18n::Backend::LazyLoad.new
+    #
+    class LazyLoad < Simple
+      # Returns whether the current locale is initialized.
+      def initialized?
+        initialized_locales[I18n.locale]
+      end
+
+      # Clean up translations and uninitialize all locales.
+      def reload!
+        @initialized_locales = nil
+        @translations = nil
+      end
+
+      protected
+
+      # Load translations from files that belong to the current locale.
+      def init_translations
+        load_translations(filenames_for_current_locale)
+        initialized_locales[I18n.locale] = true
+      end
+
+      def initialized_locales
+        @initialized_locales ||= Hash.new(false)
+      end
+
+      # Select all files from I18n load path that belong to current locale.
+      # These files must start with the locale identifier (ie. "en", "fr").
+      # or contain a path component equal to the locale identifier (ie. /locales/en/translation.yml)
+      def filenames_for_current_locale
+        I18n.load_path.flatten.select do |path|
+          (basename_starts_with_locale?(path) || path_contains_locale_identifier?(path)) &&
+          supported_extension?(path)
+        end
+      end
+
+      SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze
+
+      def supported_extension?(path)
+        SUPPORTED_EXTENSIONS.include?(File.extname(path))
+      end
+
+      def basename_starts_with_locale?(path)
+        File.basename(path).start_with?(I18n.locale.to_s)
```

Comment:
A file like `pt-BR.yml` will be loaded if the locale is `pt`, but also (again) if the locale is switched to `pt-BR`, since both match `"pt-BR.yml".start_with?(...)`. This will lead to repeat-loading which is inefficient and potentially can lead to duplicate warnings and other bugs.

It feels like we need a separator here which is not also used as part of locales themselves (i.e. something non-alphanumerical and also not `-`).

Assuming that separator is `SEPARATOR` (say `_`), then a way to do this would be:

```suggestion
        File.basename(path, ".*").split(SEPARATOR, 2)[0] == I18n.locale.to_s
```

Then `basename_starts_with_locale?("foo/pt-BR.yml")` would be true for `pt-BR` but not for `pt`. Likewise if `SEPARATOR` was `_`, then `basename_starts_with_locale?("foo/pt-BR_translation.yml")` would also be true for `pt-BR` (but not `pt`).

In the end, without some separator that is not `-`, we'll be double-loading locale files. Having a clear separator also allows us to reliably extract locales from filenames, rather than just matching against them.